cedar-lean/Cedar/Partial/Response.lean

```diff
@@ -196,4 +204,46 @@ def Response.underapproximateDeterminingPolicies (resp : Partial.Response) : Set
     -- when the decision is Unknown, nothing is guaranteed to be determining.
     Set.empty
 
+/--
+  Re-evaluate with the given substitution for unknowns, giving a new
+  `Residual`, or `none` if the residual is now `false`.
+
+  Assumes that `req` and `entities` have already been substituted.
```

Comment:
Instead of having this in a comment, can the type enforce this? (It looks like that would be easy from just this PR, but I don't have the rest of the work paged in like you do, so maybe this is completely off).

Reply:
In what way?  Having a wrapper around `Partial.Entities` that just indicates "we've performed a `subst` on this"?

An alternative is to take a `Spec.Entities` here, which would enforce that there can't be unknowns.  This points at an important(ish) `subst` design decision:
- "Partial" substitution:
    - `subst` on `Partial.Value` returns `Partial.Value`, `subst` on `Partial.Entities` returns `Partial.Entities`, etc
    - Allows "incomplete" substitutions where you only substitute some of the unknowns
    - Substitution map can be `Unknown` to `Partial.Value`, meaning you could substitute in an expression containing `unknown`s
    - `Response.reEvaluateWithSubst` has to return another `Partial.Response`
- "Full" substitution":
    - `subst` on `Partial.Value` returns `Spec.Value`, `subst` on `Partial.Entities` returns `Spec.Entities`, etc
    - Does not allow "incomplete" substitutions where you only substitute some of the unknowns
    - Substitution map has to be `Unknown` to `Spec.Value`, so you can only substitute in concrete values, not expressions containing `unknown`s
    - `Response.reEvaluateWithSubst` can return `Option Spec.Response`

This PR implements "partial" substitution, which also matches the Rust behavior better -- in Rust, `PartialResponse.reauthorize()` returns `PartialResponse` not `Response`.  But, we could also do "full" substitution, either instead or in addition.  What do you think?

Reply:
I think we want partial for maximum flexibility

Reply:
What is the use case for partial substitutions?

Reply:
The other point is that partial substitutions are more general, and we can probably implement full substitutions completely in terms of partial (plus additional preconditions like "the substitution map has a mapping for all the unknowns that appear" and "the substitution map's values contain no unknowns themselves").

Reply:
I have probably >1000 lines of proof about the definition of `subst` in this PR already, if that makes a difference.

Reply:
> we can probably implement full substitutions completely in terms of partial 

We are going to need some form of full substitution definition (either on its own or as extra constraints on partial) to state soundness, right?

I don't object to continuing with the current definitions.  My main worry would be that they will make the soundness proof harder than it would be otherwise.  The internal codebase has to do something similar to this, and it's using full substitutions.  That said, the datastructures and notions of equivalence for that proof are more complicated.  So maybe a more complex notion of substitution isn't a concern here.  That's the main thing I'd watch out for.